[DDP] Custom buffer reduction

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64515
* #64514
* __->__ #64513

Proposal: https://github.com/pytorch/pytorch/issues/63041
Support custom buffer reduction in DDP via hook. This is required for an internal training use case and could be more broadly applicable to OSS scenarios. 

Differential Revision: [D30751152](https://our.internmc.facebook.com/intern/diff/D30751152/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

rohan-varma committed:
[DDP] Custom buffer reduction

Proposal: https://github.com/pytorch/pytorch/issues/63041
Support custom buffer reduction in DDP via hook

Differential Revision: [D30751152](https://our.internmc.facebook.com/intern/diff/D30751152/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Custom buffer reduction"


Proposal: https://github.com/pytorch/pytorch/issues/63041
Support custom buffer reduction in DDP via hook. This is required for an internal training use case and could be more broadly applicable to OSS scenarios. 

Differential Revision: [D30751152](https://our.internmc.facebook.com/intern/diff/D30751152/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse agolynski @SciPioneer @H-Huang mrzzd cbalioglu gcramer23

[ghstack-poisoned]

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -1221,6 +1244,19 @@ def join_device(self):
 
+    def _register_buffer_comm_hook(
+        self,
+        state,
+        hook: callable,
```
nit: make state after hook? following up (func, args) convention?

rohan-varma replied:
Thanks for pointing this out!

I followed the same order as comm hook (state, hook): https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/distributed.py#L1224.

Should we try to keep it the same or change the order here?

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -125,6 +128,16 @@ def _dump_DDP_relevant_env_vars():
+
+@dataclass
+class _BufferCommHook:
+    buffer_comm_hook: Callable
```
nit: document how the signature of a buffer_comm_hook looks like: 

def buffer_comm_hook(state, self.name_buffers)

rohan-varma replied:
I added the documentation to the function in the PR at the top of this stack: https://github.com/pytorch/pytorch/pull/64515/files, will just leave it there if you're okay so I don't have to deal with merge conflicts.

rohan-varma committed:
Update on "[DDP] Custom buffer reduction"


Proposal: https://github.com/pytorch/pytorch/issues/63041
Support custom buffer reduction in DDP via hook. This is required for an internal training use case and could be more broadly applicable to OSS scenarios. 

Differential Revision: [D30751152](https://our.internmc.facebook.com/intern/diff/D30751152/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse agolynski @SciPioneer @H-Huang mrzzd cbalioglu gcramer23

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Custom buffer reduction"


Proposal: https://github.com/pytorch/pytorch/issues/63041
Support custom buffer reduction in DDP via hook. This is required for an internal training use case and could be more broadly applicable to OSS scenarios. 

Differential Revision: [D30751152](https://our.internmc.facebook.com/intern/diff/D30751152/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse agolynski @SciPioneer @H-Huang mrzzd cbalioglu gcramer23

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Custom buffer reduction"


Proposal: https://github.com/pytorch/pytorch/issues/63041
Support custom buffer reduction in DDP via hook. This is required for an internal training use case and could be more broadly applicable to OSS scenarios. 

Differential Revision: [D30751152](https://our.internmc.facebook.com/intern/diff/D30751152/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse agolynski @SciPioneer @H-Huang mrzzd cbalioglu gcramer23

[ghstack-poisoned]

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -887,6 +911,16 @@ def forward(self, *inputs, **kwargs):
 
+            # sync params according to location (before/after forward) user
+            # specified as part of hook, if hook was specified.
+            if (
```
moved to helper fn in the next diff

rohan-varma committed:
Update on "[DDP] Custom buffer reduction"


Proposal: https://github.com/pytorch/pytorch/issues/63041
Support custom buffer reduction in DDP via hook. This is required for an internal training use case and could be more broadly applicable to OSS scenarios. 

Differential Revision: [D30751152](https://our.internmc.facebook.com/intern/diff/D30751152/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse agolynski @SciPioneer @H-Huang mrzzd cbalioglu gcramer23

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Custom buffer reduction"


Proposal: https://github.com/pytorch/pytorch/issues/63041
Support custom buffer reduction in DDP via hook. This is required for an internal training use case and could be more broadly applicable to OSS scenarios. 

Differential Revision: [D30751152](https://our.internmc.facebook.com/intern/diff/D30751152/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse agolynski @SciPioneer @H-Huang mrzzd cbalioglu gcramer23

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Custom buffer reduction"


Proposal: https://github.com/pytorch/pytorch/issues/63041
Support custom buffer reduction in DDP via hook. This is required for an internal training use case and could be more broadly applicable to OSS scenarios. 

Differential Revision: [D30751152](https://our.internmc.facebook.com/intern/diff/D30751152/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse agolynski @SciPioneer @H-Huang mrzzd cbalioglu gcramer23

[ghstack-poisoned]
